### PR TITLE
Expose toJSON and fromJSON on AuthCredential

### PR DESCRIPTION
**Pull request by wti806**
- Exposed toJSON and fromJSON on AuthCredential.
- Changed AuthCredential from interface to abstract class as fromJSON is a static method.
- Changed the OAuthCredential constructor to take a OAuthProviderOptions object, but not expose the API to public. The constructor change is backward compatible as it still can take idToken and accessToken string.

### CODE REVIEW

File: packages/auth-types/index.d.ts [hunk 1]
```diff
@@ -246,7 +248,7 @@ export interface UserMetadata {
 
 export type Persistence = string;
 
-export interface OAuthCredential extends AuthCredential {
+export abstract class OAuthCredential extends AuthCredential {
```
Comment (bojeil-google) [inline packages/auth-types/index.d.ts]: `OAuthCredential` is not really abstract.

Comment (wti806) [reply packages/auth-types/index.d.ts]: Changed to concrete class and made the constructor private.

File: packages/auth/src/authcredential.js [hunk 2]
```diff
@@ -635,23 +635,31 @@ fireauth.OAuthProvider.prototype.getScopes = function() {
...
  * @param {?string=} opt_accessToken The optional OAuth access token.
- * @param {?string=} opt_rawNonce The optional raw nonce.
  * @return {!fireauth.AuthCredential} The Auth credential object.
  */
-fireauth.OAuthProvider.prototype.credential = function(opt_idToken,
-    opt_accessToken, opt_rawNonce) {
-  if (!opt_idToken && !opt_accessToken) {
+fireauth.OAuthProvider.prototype.credential =
+    function(optionsOrIdToken, opt_accessToken) {
+  var oauthResponse;
```
Comment (Feiyang1) [inline packages/auth/src/authcredential.js]: Nit: Can you please run prettifier through the code? The formatting is off.

Comment (wti806) [reply packages/auth/src/authcredential.js]: I believe the prettifier will be automatically triggered before pushing to canonical repo. As you can see, there are 2 commits from the prettier in this PR: 
[AUTOMATED]: Prettier Code Styling
 Do I need to manually run the prettier?

**Pull request merged**